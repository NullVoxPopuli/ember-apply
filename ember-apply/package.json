{
  "name": "ember-apply",
  "version": "2.2.1",
  "description": "automatic integration and configuration from the community",
  "repository": {
    "url": "https://github.com/NullVoxPopuli/ember-apply",
    "type": "https",
    "directory": "packages/ember-apply"
  },
  "main": "src/index.js",
  "exports": {
    ".": "./src/index.js",
    "./test-utils": "./src/test-utils.js"
  },
  "type": "module",
  "author": "NullVoxPopuli",
  "license": "MIT",
  "private": false,
  "bin": {
    "ember-apply": "src/cli/index.js"
  },
  "types": "./declarations/src/",
  "files": [
    "src/**/*",
    "declarations/**/*",
    "README.md"
  ],
  "scripts": {
    "prepack": "yarn build:types && cp ../README.md ./",
    "build:types": "tsc --build",
    "lint": "yarn lint:js",
    "lint:js": "eslint .",
    "lint:fix": "yarn lint:js --fix",
    "test": "vitest --coverage --no-watch --no-threads",
    "test:watch": "vitest --watch --no-threads"
  },
  "dependencies": {
    "@babel/core": "^7.16.7",
    "@babel/preset-env": "^7.16.8",
    "chalk": "^5.0.0",
    "ember-template-recast": "^6.1.2",
    "execa": "^6.0.0",
    "fs-extra": "^10.0.0",
    "jscodeshift": "^0.13.1",
    "latest-version": "^6.0.0",
    "ora": "^6.0.1",
    "pacote": "^13.0.0",
    "posthtml": "^0.16.5",
    "semver": "^7.3.5",
    "unified": "^10.1.1",
    "yargs": "^17.3.1",
    "yarn-workspaces-list": "^0.2.0"
  },
  "devDependencies": {
    "@nullvoxpopuli/eslint-configs": "^2.1.17",
    "@types/fs-extra": "^9.0.13",
    "@types/jscodeshift": "^0.11.3",
    "@types/node": "^17.0.8",
    "@types/pacote": "^11.1.3",
    "@types/semver": "^7.3.9",
    "@types/yargs": "^17.0.8",
    "@vitest/ui": "^0.6.0",
    "babel-eslint": "^10.1.0",
    "c8": "^7.11.0",
    "eslint": "^7.0.0",
    "eslint-plugin-decorator-position": "^4.0.1",
    "eslint-plugin-import": "^2.25.4",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^4.0.0",
    "prettier": "^2.5.1",
    "semantic-release": "^19.0.0",
    "semantic-release-monorepo": "^7.0.0",
    "typescript": "4.6.2",
    "vite": "^2.7.13",
    "vitest": "^0.5.0"
  },
  "engines": {
    "node": ">=16.0.0"
  },
  "volta": {
    "node": "16.14.0",
    "yarn": "1.22.17"
  }
}
