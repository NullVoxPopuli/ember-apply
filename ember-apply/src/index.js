// @ts-check
//
//
// Each of these re-exported modules points at an index.js file which
// defines the Public API for each namespace

export * as automation from './-private/automation/index.js';
export * as css from './-private/css/index.js';
export * as ember from './-private/ember/index.js';
export * as files from './-private/files/index.js';
export * as html from './-private/html/index.js';
export * as js from './-private/js/index.js';
export * as npm from './-private/npm/index.js';
export * as packageJson from './-private/package-json/index.js';
export * as project from './-private/project/index.js';
