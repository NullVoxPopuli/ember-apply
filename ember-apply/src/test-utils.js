import { execa } from 'execa';
import fs from 'fs/promises';
import os from 'os';
import path, { dirname } from 'path';
import { fileURLToPath } from 'url';

import { packageJson, project } from './index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

export async function newTmpDir() {
  const tmpDir = await fs.mkdtemp(
    path.join(os.tmpdir(), `ember-apply-${Date.now()}-`),
  );

  return tmpDir;
}

export async function newEmberApp() {
  let dir = await newTmpDir();

  await execa('ember', ['-v'], { cwd: dir });
  await execa('ember', ['new', 'test-app', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-app');
}

export async function newEmberAddon() {
  let dir = await newTmpDir();

  await execa('ember', ['-v'], { cwd: dir });
  await execa('ember', ['addon', 'test-addon', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-addon');
}

/**
 * @param {string} appPath the app to apply the applyable to
 * @param {string} applyablePath the path to the applyable
 */
export async function apply(appPath, applyablePath) {
  let cli = path.join(__dirname, 'cli/index.js');
  let target = applyablePath.startsWith('@')
    ? applyablePath
    : path.resolve(applyablePath);

  await execa('node', [cli, '--verbose', target], {
    cwd: appPath,
    stdio: process.env.DEBUG ? 'inherit' : undefined,
  });
}

/**
 * @param {string} appPath the path to git diff
 */
export async function diffSummary(appPath) {
  let { stdout } = await execa('git', ['diff', '--compact-summary'], {
    cwd: appPath,
  });

  return stdout;
}

/**
 * @typedef {object} DiffOptions
 * @property {boolean} [ignoreVersions]
 *
 * @param {string} appPath the path to git diff
 * @param {DiffOptions} [options]
 */
export async function diff(appPath, options = {}) {
  let { stdout } = await execa('git', ['diff', '--name-only'], {
    cwd: appPath,
  });

  let files = stdout.split('\n');

  let diff = '';

  for (let filePath of files) {
    if (!filePath) {
      continue;
    }

    if (filePath.endsWith('package.json') && options?.ignoreVersions) {
      let fileBuffer = await fs.readFile(path.join(appPath, filePath));
      let fileString = fileBuffer.toString();
      let json = JSON.parse(fileString);

      for (let section of [
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'engines',
        'volta',
      ]) {
        if (json[section]) {
          json[section] = Object.keys(json[section]);
        }
      }

      diff += `\n${filePath}\n${JSON.stringify(json, null, 2)}\n`;
      continue;
    }

    // Why is this not easier....
    // let normalCommand =
    //   `git diff ${filePath} ` +
    //   `| sed -n '/^---/!p' ` +
    //   `| sed -n '/^+++/!p' ` +
    //   `| sed -n '/^@@/!p' ` +
    //   `| sed -n '/^index /!p'`;

    let { stdout: fullDiff } = await execa(`git`, ['diff', filePath], {
      cwd: appPath,
    });

    let shortDiff = fullDiff
      .replace(/^---.+$/, '')
      .replace(/^\+\+\+.+$/, '')
      .replace(/index.+$/, '')
      .replace(/@@.+$/, '');

    diff += shortDiff + '\n';
  }

  return diff;
}

/**
 * Gets text content of the file all in one go
 * fs.readFile with the extra steps
 *
 * @param {string} filePath
 * @returns {Promise<string>} the file
 */
export async function readFile(filePath) {
  let file = await fs.readFile(filePath);

  return file.toString();
}

/**
 * Gets JSON content of the file all in one go
 * fs.readFile with the extra steps
 *
 * @param {string} filePath
 * @returns {Promise<object>} the file
 */
export async function readJson(filePath) {
  let text = await readFile(filePath);

  return JSON.parse(text);
}

/**
 * Creates a new monorepo with git in a tmp directory
 *
 * @param {string[]} workspaces list of workspace names
 */
export async function newMonorepo(workspaces) {
  let root = await newTmpDir();

  await execa('git', ['init'], { cwd: root });

  await fs.writeFile(
    path.join(root, 'package.json'),
    `{ "workspaces": [${workspaces.map((w) => `"${w}"`)}], "private": true }`,
  );
  await fs.writeFile(
    path.join(root, 'pnpm-workspace.yaml'),
    `packages:\n` + workspaces.map((w) => `- ${w}\n`).join(''),
  );

  for (let workspace of workspaces) {
    let safeName = workspace.replaceAll('/', '-');

    await fs.mkdir(path.join(root, `${workspace}`), { recursive: true });
    await fs.writeFile(
      path.join(root, `${workspace}/package.json`),
      `{ "name": "test-${safeName}" , "dependencies": {}, "version": "0.0.0" }`,
    );
  }

  return root;
}

/**
 * @param {string} root
 */
export async function readAllPackageJson(root) {
  let workspaces = await project.getWorkspaces(root);

  let manifests = [];

  // done serially, so order is consistent
  for (let workspace of workspaces) {
    let manifest = await packageJson.read(workspace);

    manifests.push(manifest);
  }

  return manifests;
}
