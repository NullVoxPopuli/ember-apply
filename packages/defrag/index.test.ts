import { apply, diff, diffSummary, newEmberApp } from 'ember-apply/test-utils';
import { describe, expect, it } from 'vitest';

import { default as unstableEmbroider } from './index.js';

describe('unstable-embroider', () => {
  it('default export exists', () => {
    expect(typeof unstableEmbroider).toEqual('function');
  });

  describe('applying to an ember app', () => {
    it('works via CLI', async () => {
      let appLocation = await newEmberApp();

      await apply(appLocation, unstableEmbroider.path);

      expect(await diffSummary(appLocation)).toMatchSnapshot();
      expect(
        await diff(appLocation, { ignoreVersions: true })
      ).toMatchSnapshot();
    });
  });
});
