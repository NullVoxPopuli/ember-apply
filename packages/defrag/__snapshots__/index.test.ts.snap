// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unstable-embroider > applying to an ember app > works via CLI 1`] = `""`;

exports[`unstable-embroider > applying to an ember app > works via CLI 2`] = `""`;
